doc/api/esm.md

```diff
@@ -1130,18 +1130,45 @@ node --experimental-top-level-await b.mjs # works
 <!-- type=misc -->
 
 To customize the default module resolution, loader hooks can optionally be
-provided via a `--experimental-loader ./loader-name.mjs` argument to Node.js.
-
+provided via `--experimental-loader ./loader-name.mjs` arguments to Node.js
 When hooks are used they only apply to ES module loading and not to any
 CommonJS modules loaded.
 
+Loaders are modules which export functions and data which can be used by
+Node.js to modify the behavior of loading ES Modules. Loaders may be "chained"
+together by passing multiple `--experimental-loader` arguments to Node.js. A
+loader module will be imported once as a singleton.
+
+A hook has access to an additional `parentLoader` argument, which contains
+functions which call the hooks of the parent loader. When calling a parent hook,
+`parentLoader` **does not** need to be passed
```

Comment:
```suggestion
`parentLoader` **does not** need to be passed.
```

What is a “parent” hook? Can we use terms like `next` like Express middleware?

Reply:
if not parent, i'd call it "previous", so idk. we've used the term parent pretty consistently throughout the last year or so of discussion and it hasn't been a problem.

Reply:
If we call it parent, then we shouldn’t call it a “chain of loaders” in the introduction imo. If we set up loaders as a “chain” then there are clearly defined terms: it starts with the first loader in the list which then calls the “next” loader in the chain and so forth.

I’m not sure which terminology leads to parent, at least not in this direction. Since the current loader calls the next loader, in terms of call stacks it would be childLoader, not parentLoader. I wouldn’t be surprised if the term “parent loader” got used and somebody understood - but thought it was talking about the opposite thing.

Reply:
in this: `node --experimental-loader a --experimental-loader b`

the resolution starts at `b`, which can choose to call to the "parent"/"previous" loader `a`.

Reply:
Wait, so the last loader to be passed is called first..? That feels *very* surprising. What kind of chain would call the thing being called to run next the "previous item"? Can you elaborate why this is different from middleware/interceptors that use the reverse terminology? Is there some precedent you're looking at?

Reply:
Well I don't think the middleware analogy helps that much because middleware step through the entire chain with each one providing modifications or additions to the context, whereas loader hooks go until the first hook that gives a result. My intuition here was that hooks refer up to a "parent" hooks when they don't know the answer, all the way up to the default resolver. I can try experimenting with your order but I will say it wasn't immediately obvious to me.

Reply:
Some other possible terms may be "baseLoader"? Or something more abstract like "delegate"?

The reason why I was thinking about the "middleware chain" a la express was because it is also a list of functions where the first is called and then optionally calls the next middleware to delegate. It gets very uncanny when looking at koa or any other interceptor middleware solutions that allow acting after the next handler is done:

```js
// Koa middleware:
app.use(async (ctx, next) => {
  // The middleware may decide not to delegate down the chain.
  if (req.path.endsWith('.coffee')) {
    return res.makeCoffee();
  }

  // Pass "arguments" to the next middleware / set up `ctx`.

  await next(); // call the next function in the chain

  // Act on the "return value" / changes to `ctx`.
});

// Loader hook:
export function resolve(specifier, parent, next) {
  if (specifier.endsWith('.coffee')) {
    return makeCoffee();
  }

  // Change the arguments for the next function

  const nextResult = await next.resolve(specifier, parent);

  // Process the result returned by the next resolve hook in the chain
}
```

There's definitely merit to different perspectives here. I can see where the parent thing comes from even though it doesn't match by intuition. It just feels like we're mixing terminology, especially if we use a very common ecosystem term ("[middleware] chain" or "[call] chain") but then use unrelated terms to talk about the elements in the chain ("parent/child"). If we want to use parent/child, we should introduce the hook collection as something that has parent/child relationships. A chain (as I understand it) doesn't really have the notion of "parent" or "child".[1]

I think my primary problem with `parent` is that it reuses the same term ("parent") as `parentURL` which also appears in this API but with an entirely different meaning. Maybe another solution would also be to change `parentURL` to `referrerURL` or `baseURL`.

[1] Unless we're talking about the chain as a list of functions and the list as a special case of a tree. But that feels very complicated.

Reply:
> loader hooks go until the first hook that gives a result

This is not what I was expecting. For example in CoffeeScript it's very common for a `.coffee` file to get transpiled into modern JavaScript by the CoffeeScript compiler, and then Babel transpiles that into some lower target. I'm used to things like Browserify where I can chain transforms like `[coffeeify, babelify]` etc. I'm expecting loader hooks to behave the same way, where e.g. `transformSource` hooks are chained, so that potentially _all_ loaded custom `transformSource` hooks get run in sequence, rather than just some or one of them.

Reply:
@GeoffreyBooth right, i mean resolve/getSource/getFormat. transformSource and getGlobalPreloadCode don't apply well to chaining and they need to behave differently (which is something I mentioned in the OP).

Reply:
Yes this is where the hooks having different models comes in. See eg RollupJS's plugin driver hook modes (https://github.com/rollup/rollup/blob/master/src/utils/PluginDriver.ts#L117) and Webpack Tapable (https://github.com/webpack/tapable).

`transformSource` is a `hookReduce` in RollupJS terminology or a `waterfall` hook in Webpack terminology.

It would be good to ensure it behaves this way.

Reply:
One way to chain transform without losing information would be to retain the original source result of `getSource` in the `context`.